Remote indexes

guillemcortes opened the pull request:
Addresses #155 , #156 and #159

guillemcortes committed:
Remote indexes (pending tests)

guillemcortes commented:
TODO
 The goal is to mimic mirdata remote index management since it also allows version control. The only downside is that it requires to have an "index sample" for testing (like the .wav files). Check the mirdata PR: https://github.com/mir-dataset-loaders/mirdata/pull/496
 
 - [x] `Index` class in core.py
 - [x] define indexes in all loaders as INDEXES (for reference: https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/datasets/acousticbrainz_genre.py#L63)
 - [x] Modify `Dataset` class to match [mirdat's](https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/core.py#L74)
     - [x] remove `custom_index_path` attribute of `Dataset` class.
     - [x] Add `version`
     - [x] Add `indexes`
 - [x] Modify general `tests`
 - [x] Create sample json indexes for all soundata indexes.
 - [x] downloader(),... add remote index
 - [x] Augment this list after reviewing [mirdata's PR 496](https://github.com/mir-dataset-loaders/mirdata/pull/496)
     - [x] Modify example.py and test_example.py from the documentation
     - [x] Soundata __init__.py
     - [x] Check #TODO comments in guillemcortes' commits 
 - [x] Validate index!
 - [x] Consider remote indexes as the only option. Remove conditions
 - [x] Fix index storage path
 - [x] DOCUMENT EVERYTHING NEW :)
     - [x] As a contributor, where to upload the index
     - [x] As a contributor, creating sample index, where to put it, how to test, etc.
     - [x] As a user, how to download and use index
 - [x] test that all `_sample` indexes version is "sample"
 - [x] Remove package data in `setup.py`
 - [x] Remove all indexes from soundata/datasets/indexes
 - [x] Do we need any extra/new test?
 - [x] Delete "working with remote indexes" from docs
 - [x] Document Sample Indexes inside "Add tests"
 - [x] Document how to create the Sample index and how to define it, example of INDEXES variable in the loader...
 - [x] Update docs/source/contributing_examples with the remote indexes

guillemcortes committed:
Add class Index at core.py and downloader()

guillemcortes committed:
Modify example.py and test_example.py

guillemcortes committed:
Modify soundata/__init__.py

guillemcortes committed:
Missing version arg in docstring

guillemcortes committed:
Minor change in docstring

guillemcortes committed:
modify conftest

guillemcortes committed:
Modify general tests

guillemcortes committed:
Black

guillemcortes commented:
If we follow mirdata, they store sample indexes in https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/datasets/indexes/acousticbrainz_genre_index_sample.json `mirdata/datasets/indexes/index_sample.json` since we plan to have all indexes hosted in Zenodo, I think it makes more sense if the sample indexes that are only used for testing are stored in `tests/indexes/`. What do you think? @genisplaja @tanmayy24

guillemcortes committed:
Loader bioacoustic test

guillemcortes committed:
Missing version arg in init

guillemcortes committed:
dcase_birdVox20k example

guillemcortes committed:
WIP fixing core.py and test_core.py

genisplaja commented on soundata/datasets/indexes/dcase_birdVox20k_index_1.0_sample.json:
```diff
@@ -0,0 +1,11 @@
+{
+    "version": "1.0",
```
`1.0` --> `sample`

genisplaja committed:
multiple updates in source and tests for remote indexes

guillemcortes committed:
Adapt first batch of dataloaders and tests

genisplaja committed:
untracked work

genisplaja committed:
Merge remote-tracking branch 'origin/guillem/remote-indexes' into guillem/remote-indexes

genisplaja committed:
documentation on remote indexes

guillemcortes committed:
another batch of loaders fixed

guillemcortes committed:
Merge branch 'main' into guillem/remote-indexes

guillemcortes committed:
Last batch of loaders

guillemcortes committed:
Some fixes

guillemcortes committed:
tests passing

guillemcortes committed:
Black formatting

guillemcortes committed:
Merge branch 'main' into guillem/remote-indexes

guillemcortes committed:
Remove package data

guillemcortes committed:
Test Index()

guillemcortes committed:
Increase test coverage

guillemcortes committed:
Black :)

guillemcortes committed:
Test download utils

guillemcortes committed:
Remove indexes

guillemcortes committed:
Indexes folder readme

guillemcortes committed:
add version=test to dataset in test

guillemcortes committed:
Docs update

guillemcortes committed:
fix in tests

guillemcortes committed:
fixing index path in marco dataset

guillemcortes committed:
Update index path in multiple datasets

guillemcortes committed:
Revert dataset version change in test_core

guillemcortes committed:
-.-

guillemcortes committed:
test sample index version

guillemcortes committed:
Change index version to sample for sample indexes

guillemcortes commented:
Hi @hadware, @hagenw,

We would like to invite you to review this PR that addresses the issues you raised during JOSS review: #155 #156 and #159 .

Summarizing the PR:
- Indexes will be stored at https://zenodo.org/communities/audio-data-loaders/ from now on.
- Sample indexes for testing will have to be created and packaged with Soundata (similarly to what we do with audio files for testing)
- Created an Index class that supports different versions of a dataset. (We adopted the same approach as in [mirdata](https://github.com/mir-dataset-loaders/mirdata/pull/496).
- Document the new process for both users and contributors.
- Add tests accordingly :)

Any comments/suggestions/concerns you might have would be really helpful to us.

rythmm24 committed:
addition to remove test files from pip

hadware commented:
Perfect! This appears to perfectly address these issues.

I would advise cleaning up the git tree after you've merged your PR (using https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository , this _will_ required a force-push at some point, as it rewrites the diff tree).

genisplaja committed:
Tanmay's comments first iteration

genisplaja commented on docs/source/contributing.rst:
```diff
@@ -282,6 +284,33 @@ See an example of how an index should look like:
+To prevent an the exponential growth in size of the library, we are storing all indexes in an online repository, while removing these from the repository itself.
+For that task, we use Zenodo.
+To use a dataloader, users may retrieve the index running the ``dataset.download()`` function that is also used to download the dataset.
+To download only the index, you may run ``.download(["index"])``. The index will be automatically downloaded and stored in the expected folder in Soundata.
```
I think that now the message should be improved. Can you please check @tanmayy24?

genisplaja replied:
indeed. Fixed now :) Maybe it's a little bit overkill to use `if` `elif` `else` but that's the easiest way I found to log more meaningful messages here.

genisplaja committed:
improving downloading logs and index documentation

guillemcortes commented:
This looks good! Should we merge? @genisplaja @tanmayy24

hagenw commented on docs/source/contributing.rst:
```diff
@@ -282,6 +284,37 @@ See an example of how an index should look like:
+----------------------------------------------
+
+To prevent an the exponential growth in size of the library, we are storing all indexes in an online repository, while removing these from the repository itself.
+For that task, we use Zenodo.
```
This sentence seems to make sense only for already existing indices. I guess new ones are never added to the repo in the first place:

```suggestion
We store all indexes in an online repository on Zenodo.
```

genisplaja replied:
hey @hagenw, done! Thanks for the suggestion.

hagenw commented on docs/source/tutorial.rst:
```diff
@@ -42,11 +42,18 @@ To use a loader, (for example, ``urbansound8k``) you need to initialize it by ca
-version of the dataset (when available). When initializing a dataset it is important to correctly set up the directory
+All dataset loaders in soundata have a ``download()`` function that allows the user to download:
+
+* The :ref:`canonical <faq>` version of the dataset (when available).
```
When is no canonical version of the dataset available?
When there are more than one version?
What happens then, will the latest version be downloaded automatically, or is the user forced to specify a version?

genisplaja replied:
Ok, thanks! Added documentation and examples about these issues. To better support the use of different versions, I did add a `list_dataset_versions()` function in `soundata.__init__` which should help users get the available versions before initializing the data loaders. This is documented as well. This is all in my last commit.

Does that address the issue @hagenw? Please @tanmayy24 and @guillemcortes also take a look to see if that sounds good to you.

hagenw replied:
Thanks, for updating the docs accordingly.
Looks fine to me, I just added a few small comments.

hagenw commented:
Nice to see, that this also helps addressing the problem of dataset versioning.

If the merge request should be ready for merging, I would recommend to remove `[WIP]` from its title ;)

genisplaja committed:
initializing docs, listing versions, typos

genisplaja committed:
test list_dataset_versions()

hagenw commented on docs/source/tutorial.rst:
```diff
@@ -42,14 +42,32 @@ To use a loader, (for example, ``urbansound8k``) you need to initialize it by ca
 You can indicate where the data would be stored and access by passing a path to ``data_home``, as explained below. Now ``us8k`` is a ``Dataset``
 object containing common methods, described in the following.
 
+Some datasets have more that one version available. Soundata can handle more than one version if included in the dataloader.
```
```suggestion
Some datasets have more than one version available. Soundata can handle more than one version if included in the dataloader.
```

BTW, I would highly recommend to use semantic linefeeds in the future, when writing documentation (sorry, can only share as an image):

![image](https://github.com/soundata/soundata/assets/173624/99648f70-b011-4fa6-abfe-b6a5d96d7b5b)

genisplaja replied:
I see. Indeed, thanks for the suggestion. Actually, longer lines make editing more difficult as well, since you cannot read the entire thing and you have to move left and right a lot. Will take into account for sure!

hagenw commented on docs/source/tutorial.rst:
```diff
@@ -42,14 +42,32 @@ To use a loader, (for example, ``urbansound8k``) you need to initialize it by ca
+    dataset = soundata.initialize('urbansound8k', data_home='choose_where_data_live', version="1.0")
+
+By default, version is set to ``'default'``, which routes to the latest version available in the dataloader. 
+You may run ``soundata.list_dataset_versions('urbansound8k')`` to see all available versions of the dataset that are available through its dataloder.
```
```suggestion
You may run ``soundata.list_dataset_versions('urbansound8k')`` to see all available versions of the dataset, that are available through its dataloder.
```

genisplaja replied:
Ok! I am waiting until we resolve the versioning issue here: https://github.com/soundata/soundata/pull/169#discussion_r1616019129, since it'll most likely make this line change again :)

hagenw commented on docs/source/tutorial.rst:
```diff
@@ -42,14 +42,32 @@ To use a loader, (for example, ``urbansound8k``) you need to initialize it by ca
+    import soundata
+    dataset = soundata.initialize('urbansound8k', data_home='choose_where_data_live', version="1.0")
+
+By default, version is set to ``'default'``, which routes to the latest version available in the dataloader. 
```
Latest refers to the latest published version, not the on with the highest version number, correct?
E.g. assume you have a dataset with versions 1.0 and 2.0, and then you publish a fix for 1.0 as 1.1, then I guess `latest` refers to 1.1.

genisplaja replied:
That's right @hagenw. I was trying to mention that we point to the _highest_ version. Therefore, in the example you posed, the default would still point to 2.0. I actually wonder if it is actually right to assume that we should be pointing to the highest version, probably not. It will most likely depend on each case. 

One thing that we could do, is letting the data loader creator to decide which version should be defaulted, and then update the `list_dataset_versions()` so that instead of returning the list of version numbers, displays a message like:

```
Dataset <dataset-name> has the following versions available: ["1.0", "1.1", "2.0"].
Default version for <dataset-name> is "2.0".
```
Does that sound good? Of course we would document these change as such.

genisplaja replied:
Nonetheless, if we want a criteria for that issue we could ask the contributors to default to the highest version unless there is a motivation to point to a different one (maybe the highest one is really recent and most of the works are based on the old one, etc.)

hagenw replied:
I think allowing to set a default version by the dataset maintainer makes sense, otherwise you cannot choose a default with a lower version number.

genisplaja committed:
that->than in tutorial docs

Co-authored-by: Hagen Wierstorf <hagen.wierstorf@posteo.de>

genisplaja committed:
multiple documentation fixes for dataset versioning

genisplaja committed:
merging

genisplaja committed:
fix list versions test

genisplaja committed:
black on tests

genisplaja committed:
minor doc improvements

guillemcortes committed:
Doc improvements

guillemcortes committed:
Update docs

guillemcortes committed:
Pull request info update

guillemcortes committed:
Minor changes documentation

guillemcortes committed:
Increase test coverage

guillemcortes committed:
black

guillemcortes committed:
Improve installing soundata for dev documentation

guillemcortes committed:
Minor change in docs

genisplaja committed:
add warning to add index checksum before running test_full_dataset

faroit commented:
wow, this is cool
